api/src/test/java/org/apache/iceberg/expressions/TestEvaluator.java

```diff
@@ -362,4 +372,198 @@ public void testCharSeqValue() {
+        in("x", Long.MAX_VALUE, Integer.MAX_VALUE, Long.MIN_VALUE));
+    Assert.assertTrue("Integer.MAX_VALUE in [Integer.MAX_VALUE] => true",
+        intSetEvaluator.eval(TestHelpers.Row.of(Integer.MAX_VALUE, 7.0, null)));
+    Assert.assertFalse("6 in [Integer.MAX_VALUE]  => false",
+        intSetEvaluator.eval(TestHelpers.Row.of(6, 6.8, null)));
+
+    Evaluator integerEvaluator = new Evaluator(STRUCT, in("y", 7, 8, 9.1));
+    Assert.assertTrue("7.0 in [7, 8, 9.1] => true",
+        integerEvaluator.eval(TestHelpers.Row.of(0, 7.0, null)));
+    Assert.assertTrue("9.1 in [7, 8, 9.1] => true",
+        integerEvaluator.eval(TestHelpers.Row.of(7, 9.1, null)));
+    Assert.assertFalse("6.8 in [7, 8, 9.1]  => false",
+        integerEvaluator.eval(TestHelpers.Row.of(6, 6.8, null)));
+
+    Evaluator structEvaluator = new Evaluator(STRUCT, in("s1.s2.s3.s4.i", 7, 8, 9));
+    Assert.assertTrue("7 in [7, 8, 9] => true",
+            structEvaluator.eval(TestHelpers.Row.of(7, 8, null,
+                    TestHelpers.Row.of(
+                            TestHelpers.Row.of(
+                                    TestHelpers.Row.of(
+                                            TestHelpers.Row.of(7)))))));
+    Assert.assertFalse("6 in [7, 8, 9]  => false",
+            structEvaluator.eval(TestHelpers.Row.of(6, 8, null,
+                    TestHelpers.Row.of(
+                            TestHelpers.Row.of(
+                                    TestHelpers.Row.of(
+                                            TestHelpers.Row.of(6)))))));
+
+    StructType charSeqStruct = StructType.of(required(34, "s", Types.StringType.get()));
+    Evaluator charSeqEvaluator = new Evaluator(charSeqStruct, in("s", "abc", "abd", "abc"));
+    Assert.assertTrue("utf8(abc) in [string(abc), string(abd)] => true",
+            charSeqEvaluator.eval(TestHelpers.Row.of(new Utf8("abc"))));
+    Assert.assertFalse("utf8(abcd) in [string(abc), string(abd)] => false",
+            charSeqEvaluator.eval(TestHelpers.Row.of(new Utf8("abcd"))));
+  }
+
+  @Test
+  public void testInExceptions() {
+    TestHelpers.assertThrows(
+        "Throw exception if value is null",
+        NullPointerException.class,
+        "Cannot create expression literal from null",
+        () -> in("x", (Literal) null));
+
+    TestHelpers.assertThrows(
+        "Throw exception if value is null",
+        NullPointerException.class,
+        "Values cannot be null for IN predicate",
+        () -> in("x", (Collection<?>) null));
+
+    TestHelpers.assertThrows(
+        "Throw exception if calling literal() for IN predicate",
+        IllegalArgumentException.class,
+        "IN predicate cannot return a literal",
+        () -> in("x", 5, 6).literal());
+
+    TestHelpers.assertThrows(
+        "Throw exception if any value in the input is null",
+        NullPointerException.class,
+        "Cannot create expression literal from null",
+        () -> in("x", 1, 2, null));
```

Comment:
We will need to be very careful handling nulls as they have special meaning in SQL and query engines can also push down `IN` with nulls.

Reply:
Currently, `null` has to be handled separately using `isNull`. 
If `null` support is needed, we might handle it by replacing it with a special `NULL` value (e.g. enum `NULL`).